### PR TITLE
Use environment COQLIB to guess -coqlib if defined.

### LINKED ISSUE #5975: Coq 8.7.0 test-suite fails (when Coq not configured -local and not installed)
I'm not sure what's up with `Error: cannot guess a path for Coq libraries; please use -coqlib option`.  See [this launchpad log](https://launchpadlibrarian.net/341301069/buildlog_ubuntu-precise-amd64.coq-8.7.0_8.7.0-1~precise~ppa79_BUILDING.txt.gz) for more details.
```
make[5]: Entering directory `/<<PKGBUILDDIR>>/test-suite'
Makefile:177: warning: undefined variable `TRAVIS'
logs/misc/printers.log
==========> TESTING misc/printers.sh <==========
Error during initialization: 
Error: cannot guess a path for Coq libraries; please use -coqlib option
0m0.000000s 0m0.000000s
0m0.050000s 0m0.000000s
==========> FAILURE <==========
    misc/printers.sh...Error!

logs/coq-makefile/uninstall1.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/uninstall1/run.sh...Error!

logs/coq-makefile/findlib-package.log
../template/init.sh: line 8: git: command not found
make[5]: Entering directory `/<<PKGBUILDDIR>>/test-suite/coq-makefile/findlib-package/findlib/foo'
make[5]: Leaving directory `/<<PKGBUILDDIR>>/test-suite/coq-makefile/findlib-package/findlib/foo'
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/findlib-package/run.sh...Error!

logs/coq-makefile/arg.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/arg/run.sh...Error!

logs/coq-makefile/extend-subdirs.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/extend-subdirs/run.sh...Error!

logs/coq-makefile/only.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/only/run.sh...Error!

logs/coq-makefile/coqdoc1.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/coqdoc1/run.sh...Error!

logs/coq-makefile/multiroot.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/multiroot/run.sh...Error!

logs/coq-makefile/latex1.log
/usr/bin/pdflatex
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/latex1/run.sh...Error!

logs/coq-makefile/mlpack1.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/mlpack1/run.sh...Error!

logs/coq-makefile/plugin3.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/plugin3/run.sh...Error!

logs/coq-makefile/merlin1.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/merlin1/run.sh...Error!

logs/coq-makefile/coqdoc2.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/coqdoc2/run.sh...Error!

logs/coq-makefile/uninstall2.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/uninstall2/run.sh...Error!

logs/coq-makefile/plugin2.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/plugin2/run.sh...Error!

logs/coq-makefile/plugin1.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/plugin1/run.sh...Error!

logs/coq-makefile/compat-subdirs.log
../template/init.sh: line 8: git: command not found
Warning: Subdirectories are deprecated.
Warning: Use double colon rules in Makefile.local.

Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/compat-subdirs/run.sh...Error!

logs/coq-makefile/mlpack2.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/mlpack2/run.sh...Error!

logs/coq-makefile/validate1.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/validate1/run.sh...Error!

logs/coq-makefile/timing.log
../template/init.sh: line 8: git: command not found
Error: cannot guess a path for Coq libraries; please use -coqlib option
==========> FAILURE <==========
    coq-makefile/timing/run.sh...Error!

FAILURES
    misc/printers.sh...Error!
    coq-makefile/arg/run.sh...Error!
    coq-makefile/compat-subdirs/run.sh...Error!
    coq-makefile/coqdoc1/run.sh...Error!
    coq-makefile/coqdoc2/run.sh...Error!
    coq-makefile/extend-subdirs/run.sh...Error!
    coq-makefile/findlib-package/run.sh...Error!
    coq-makefile/latex1/run.sh...Error!
    coq-makefile/merlin1/run.sh...Error!
    coq-makefile/mlpack1/run.sh...Error!
    coq-makefile/mlpack2/run.sh...Error!
    coq-makefile/multiroot/run.sh...Error!
    coq-makefile/only/run.sh...Error!
    coq-makefile/plugin1/run.sh...Error!
    coq-makefile/plugin2/run.sh...Error!
    coq-makefile/plugin3/run.sh...Error!
    coq-makefile/timing/run.sh...Error!
    coq-makefile/uninstall1/run.sh...Error!
    coq-makefile/uninstall2/run.sh...Error!
    coq-makefile/validate1/run.sh...Error!
make[5]: *** [report] Error 1
make[5]: Leaving directory `/<<PKGBUILDDIR>>/test-suite'
make[4]: *** [all] Error 2
make[4]: Leaving directory `/<<PKGBUILDDIR>>/test-suite'
make[3]: *** [test-suite] Error 2
make[3]: Leaving directory `/<<PKGBUILDDIR>>'
make[2]: *** [submake] Error 2
make[2]: Leaving directory `/<<PKGBUILDDIR>>'
make[1]: *** [override_dh_auto_test] Error 2
make[1]: Leaving directory `/<<PKGBUILDDIR>>'
make: *** [build] Error 2
```

### PR DESCRIPTION
**Pull request by SkySkimmer**
I don't know if it makes sense to backport this.

- [ ] Corresponding documentation was added / updated.
- [ ] Entry added in CHANGES.

### CODE REVIEW

File: tools/CoqMakefile.in [hunk 1]
```diff
@@ -177,10 +179,16 @@ COQEXTRAFLAGS?=
 COQCHKEXTRAFLAGS?=
 COQDOCEXTRAFLAGS?=
 
+ifneq ($(EXPLICIT_COQLIB),)
```
Comment (SkySkimmer) [inline tools/CoqMakefile.in]: I use `EXPLICIT_COQLIB` to pass -coqlib to coqc only if it was passed to coq_makefile / in the _CoqProject, but we could just always pass it.

**Review by ejgallego — COMMENTED** [review]
IMHV this is making more messy an already messy situation.

Why does properly setting `COQPATH` doesn't work in this case?

**Comment by ejgallego** [comment]
https://coq.inria.fr/refman/practical-tools/coq-commands.html#by-environment-variables

**Comment by ejgallego** [comment]
See also some discusion in https://github.com/coq/coq/wiki/CoqPaths . I think this PR should happen after the basic coqfind implementation arrives, which will be shortly, then `coqfind` will provide a unified API for locating the library and most of the things in this PR should not be necessary here IMO.

**Comment by gares** [comment]
I don't understand what problem this PR tries to solve.

**Comment by ejgallego** [comment]
> This is a promise you often makes but how reliable is this statement? And what does "shortly" mean? This is a 8.7 regression that hasn't been fixed so far and prevents (for instance), running the test-suite with nix-build.

I am a bit with @gares I am not sure the problem is worth solving, why is Nix failing here? It should install coq before tying to run the test-suite.

**Comment by ejgallego** [comment]
My point is that it is the `local` flag which is severly broken as having two file layouts for build artifacts is the biggest pain on earth. I have found this problem often and thus I'd like to abstract such layout differences first, then eliminate them.

**Review by gares — DISMISSED** [review]
Check if `export COQLIB=....` fixes the problem.
If it does not, fix the handling of `COQLIB` in `envars`.

**Comment by gares** [comment]
I guess the other invocation of `coqtop` in the test suite receive `-coqlib` option then, I've to leave now, no time to check. They should too read COQLIB, imho. E.g. if in `misc/` you call `coqwhatever` you have the same problem of `coq_makefile` not being passed `-coqlib`, while an env variable is inherited.

**Comment by SkySkimmer** [comment]
>E.g. if in misc/ you call coqwhatever you have the same problem of coq_makefile not being passed -coqlib, while an env variable is inherited.

https://github.com/coq/coq/blob/927b3826e645d428ebe3d8c6599c1f9e2bf79d46/test-suite/Makefile#L35-L39
https://github.com/coq/coq/blob/927b3826e645d428ebe3d8c6599c1f9e2bf79d46/test-suite/Makefile#L466-L469
https://github.com/coq/coq/blob/927b3826e645d428ebe3d8c6599c1f9e2bf79d46/test-suite/misc/coqc_dash_o.sh#L9

**Comment by ejgallego** [comment]
By installing I mean installing in it the build layout that is `debian/tmp` or whatever the packager chose. In that situation coqtop should locate the lib

**Comment by ejgallego** [comment]
> LOL certainly not, who does this?

That's standard procedure in most software and for a good reason.

**Comment by SkySkimmer** [comment]
COQLIB is only used by coq_makefile in the makefile, and not as an 
environment variable. coqc/coqdep/etc don't look at it.

Gaëtan Gilbert

**Comment by gares** [comment]
> COQLIB is only used by coq_makefile in the makefile, and not as an environment variable. coqc/coqdep/etc don't look at it.

Well, that's the root of all evil then (coqc calls coqtop, so it does not count)

**Comment by SkySkimmer** [comment]
New version using COQLIB.

File: lib/system.ml [hunk 2]
```diff
@@ -302,7 +302,7 @@ let with_time ~batch f x =
     raise e
 
 let get_toplevel_path top =
-  let dir = Filename.dirname Sys.argv.(0) in
```
Comment (SkySkimmer) [inline lib/system.ml]: This was broken when using Coq from PATH.

File: lib/system.ml [hunk 3]
```diff
@@ -302,7 +302,7 @@ let with_time ~batch f x =
...
+  let dir = Filename.dirname Sys.executable_name in
```
Comment (ejgallego) [inline lib/system.ml]: Uh, have you tested this with bytecode builds of all the utilities, including coqide + workers?

File: lib/system.ml (hunk 2 above)
Comment (ejgallego) [reply lib/system.ml]: It works like this [getting the absolute path] on purpose.

Comment (SkySkimmer) [reply lib/system.ml]: argv.(0) is not the absolute path.

File: lib/system.ml (hunk 3 above)
Comment (SkySkimmer) [reply lib/system.ml]: >Uh, have you tested this with bytecode builds of all the utilities, including coqide + workers?

I don't know how to do that.

Comment (ejgallego) [reply lib/system.ml]: Build a byte-only version of Coq and use `coqide.byte` , but even with the current config `coqide.byte` this to call `coqidetop.byte`, but that's always a debate given that client/servers are compatible in that sense.

Comment (SkySkimmer) [reply lib/system.ml]: >even with the current config coqide.byte this to call coqidetop.byte

What? This sentence is ill-formed.

Comment (ejgallego) [reply lib/system.ml]: `coqide.byte` tries to call `coqidetop.byte`, so I mean you usually don't need to ensure you are in a byte-only build.

Comment (ejgallego) [reply lib/system.ml]: But `Sys.executable_name` I think tends to be `ocaml` for bytecode binaries, so beware of that.

Comment (SkySkimmer) [reply lib/system.ml]:
```
gaetan@debian-hibiki ~/d/c/coq> bin/coqide.byte &
gaetan@debian-hibiki ~/d/c/coq> ps aux | grep coq
gaetan   17054  3.4  0.5 448760 60196 pts/4    Sl   15:29   0:00 /home/gaetan/.opam/4.06.0/bin/ocamlrun bin/coqide.byte
gaetan   17061  1.0  0.4 712772 55672 pts/4    Sl   15:29   0:00 /home/gaetan/.opam/4.06.0/bin/ocamlrun bin/coqidetop.byte -main-channel stdfds -control-channel 127.0.0.1:36095:39245 --xml_format=Ppcmds -async-proofs on
gaetan   17065  1.1  0.4 622572 55376 pts/4    Sl   15:29   0:00 /home/gaetan/.opam/4.06.0/bin/ocamlrun bin/coqproofworker.byte -main-channel stdfds -control-channel 127.0.0.1:54771:47601 --xml_format=Ppcmds -worker-id proofworker:0 -async-proofs-worker-priority low
gaetan   17094  0.0  0.0  13948  1028 pts/4    S+   15:29   0:00 grep --color=auto coq

```

Comment (ejgallego) [reply lib/system.ml]: Umm, if you are getting the directory where `ocamlrun` is, how can this work? Umm... I need to have a look.

Comment (SkySkimmer) [reply lib/system.ml]: If you `let () = Printf.printf "%s\n" Sys.executable_name` it prints
```
bin/coqidetop.byte
Welcome to Coq 8.8+alpha (May 2018)
bin/coqproofworker.byte
bin/coqide.byte
```
ocamlrun is just ps being well informed.

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -112,7 +112,7 @@ before_script:
     # careful with the ending /
     - BIN=$(readlink -f ../_install_ci/bin)/
     - LIB=$(readlink -f ../_install_ci/lib/coq)/
-    - make -j "$NJOBS" BIN="$BIN" LIB="$LIB" all
+    - make -j "$NJOBS" COQBIN="$BIN" COQLIB="$LIB" all
```
Comment (SkySkimmer) [inline .gitlab-ci.yml]: Almost forgot to do this change ^^

File: lib/system.ml (hunk 3 above)
Comment (ejgallego) [reply lib/system.ml]: Umm, that used not to be the case in the past, let me check in some switches.

**Review by ejgallego — COMMENTED** [review]
It doesn't make a lot of sense to me to have both `COQLIB` and `COQPATH`, so I suggest we discuss this in person next week.

**Comment by SkySkimmer** [comment]
>It doesn't make a lot of sense to me to have both COQLIB and COQPATH, so I suggest we discuss this in person next week.

We forgot about this, what now?

**Comment by ejgallego** [comment]
So my idea is indeed to have a single `COQPATH` variable, and a very simple `coqfind` library that can be used to locate libs.

- Clients do call `Coqfind.predicate "stdlib" LoadPath` and they get the corresponding flags.
- `coqfind` looks for `coqlib` files in `COQPATH` in the same way than `ocamlfind` does. It can also understand `OCAMLPATH` as to easy so particular setups.

I am positive I will have a prototype before the summer, so how to proceed, I don't know.

**Comment by ejgallego** [comment]
> So not blocking to this PR.

Well, this is introducing a new undocumented variable that modifies key parts of the system, so we either:
a) make it official
b) do it in a different way.

I think the special status of coqlib is a mistake so certainly I am not supporting a) and creating even more backwards compatiblity problems when we want to adopt a better solution.

**Comment by SkySkimmer** [comment]
I don't have a plan for this PR, if someone else wants to continue feel free.

**Pull request closed without merging**